src/reducers/moveThought.ts

```diff
@@ -158,39 +41,118 @@ const moveThought = (state: State, { oldPath, newPath, offset, skipRerank, newRa
+  const childrenOfDestination = getChildrenRankedById(state, destinationThoughtId)
+
+  /**
+   * Find duplicate thought.
+   */
+  const duplicateSubthought = () =>
+    childrenOfDestination.find(child => normalizeThought(child.value) === normalizeThought(sourceThought.value))
+
+  // if thought is being moved to the same context that is not a duplicate case
+  const duplicateThought = !sameContext ? duplicateSubthought() : undefined
+
+  const isPendingMerge = duplicateThought && (sourceThought.pending || duplicateThought.pending)
+
+  const newStateAfterMerge =
+    !!duplicateThought &&
+    !isPendingMerge &&
+    mergeThoughts(state, {
+      sourceThoughtPath,
+      targetThoughtPath: appendToPath(destinationThoughtPath, duplicateThought.id),
+    })
+
+  const destinationContext = pathToContext(state, destinationThoughtPath)
+
+  const isArchived = destinationContext?.indexOf('=archive') !== -1
+
+  // if move is used for archive then update the archived field to latest timestamp
+  const archived = isArchived ? timestamp() : destinationThought.archived
+
+  const contextIndexUpdates = newStateAfterMerge
+    ? {}
+    : {
+        // if moved within same context we don't need to remove and add source thought from source parent to the destination
+        ...(!sameContext
+          ? {
+              // remove source thought from the previous source parent children array
+              [sourceParentThought.id]: {
+                ...sourceParentThought,
+                children: sourceParentThought.children.filter(thoughtId => thoughtId !== sourceThought.id),
+                lastUpdated: timestamp(),
+                updatedBy: getSessionId(),
+              },
+              // add source thought to the destination thought children array
+              [destinationThought.id]: {
+                ...destinationThought,
+                children: [...destinationThought.children, sourceThought.id],
+                lastUpdated: timestamp(),
+                updatedBy: getSessionId(),
+              },
+            }
+          : {}),
+        // update source thought parent id, rank and other stuffs
+        [sourceThought.id]: {
+          ...sourceThought,
+          parentId: destinationThought.id,
+          rank: newRank,
+          archived,
+          lastUpdated: timestamp(),
+          updatedBy: getSessionId(),
+        },
+      }
+
   return reducerFlow([
     state => ({
       ...state,
-      contextViews: contextViewsNew,
       cursor: newCursorPath,
       ...(offset != null ? { cursorOffset: offset } : null),
     }),
-
     // update thoughts
-    updateThoughts({
-      contextIndexUpdates,
-      thoughtIndexUpdates: {},
-      recentlyEdited,
-      // load the children of the conflicted path if it's pending
-      pendingPulls: isNewContextPending ? [{ path: newPath }] : [],
-    }),
+    newStateAfterMerge
+      ? null
+      : updateThoughts({
```

Comment:
Similarly, can you add a comment here explaining what is being updated and why it only happens if `newStateAfterMerge` is null?